symbolic-base/src/ZkFold/Symbolic/Data/Ord.hs

```diff
@@ -1,158 +1,201 @@
+  default (>=)
+    :: (Ord (BooleanOf a), OrderingOf (BooleanOf a) ~ OrderingOf a)
+    => a -> a -> BooleanOf a
+  x >= y = ordering false true true (compare x y)
+  max, min :: a -> a -> a
+  max x y = ordering y x x (compare x y)
+  min x y = ordering x x y (compare x y)
+
+instance Ord Natural where
+  type OrderingOf Natural = Prelude.Ordering
+  ordering x y z = \case
+    Prelude.LT -> x; Prelude.EQ -> y; Prelude.GT -> z
+  compare = Prelude.compare
+  (>) = (Prelude.>)
+  (>=) = (Prelude.>=)
+  (<) = (Prelude.<)
+  (<=) = (Prelude.<=)
+  min = Prelude.min
+  max = Prelude.max
+instance Ord Prelude.Bool where
+  type OrderingOf Prelude.Bool = Prelude.Ordering
+  ordering x y z = \case
+    Prelude.LT -> x; Prelude.EQ -> y; Prelude.GT -> z
+  compare = Prelude.compare
+  (>) = (Prelude.>)
+  (>=) = (Prelude.>=)
+  (<) = (Prelude.<)
+  (<=) = (Prelude.<=)
+  min = Prelude.min
+  max = Prelude.max
+instance Ord Prelude.String where
+  type OrderingOf Prelude.String = Prelude.Ordering
+  ordering x y z = \case
+    Prelude.LT -> x; Prelude.EQ -> y; Prelude.GT -> z
+  compare = Prelude.compare
+  (>) = (Prelude.>)
+  (>=) = (Prelude.>=)
+  (<) = (Prelude.<)
+  (<=) = (Prelude.<=)
+  min = Prelude.min
+  max = Prelude.max
+instance KnownNat n => Ord (Zp n) where
+  type OrderingOf (Zp n) = Prelude.Ordering
+  ordering x y z = \case
+    Prelude.LT -> x; Prelude.EQ -> y; Prelude.GT -> z
+  compare = Prelude.compare
+  (>) = (Prelude.>)
+  (>=) = (Prelude.>=)
+  (<) = (Prelude.<)
+  (<=) = (Prelude.<=)
+  min = Prelude.min
+  max = Prelude.max
+
+newtype Ordering c = Ordering (c Par1)
+  deriving (Generic)
+deriving instance NFData (c Par1) => NFData (Ordering c)
+deriving instance Show (c Par1) => Show (Ordering c)
+deriving newtype instance Symbolic c => Conditional (Bool c) (Ordering c)
+deriving newtype instance Symbolic c => Eq (Ordering c)
+instance Symbolic c => SymbolicData (Ordering c)
+instance Symbolic c => Semigroup (Ordering c) where
+  Ordering o1 <> Ordering o2 = Ordering $ fromCircuit2F o1 o2 $
+    \(Par1 v1) (Par1 v2) -> Par1 <$>
+      newAssigned (\x -> let x1 = x v1; x2 = x v2 in x1 * x1 * (x1 - x2) + x2)
+instance Symbolic c => Monoid (Ordering c) where
+  mempty = eq
+instance Symbolic c => IsOrdering (Ordering c) where
+  lt = Ordering $ embed (Par1 (negate one))
+  eq = Ordering $ embed (Par1 zero)
+  gt = Ordering $ embed (Par1 one)
+
+instance (Symbolic c, LayoutFunctor f) => Ord (c f) where
+  type OrderingOf (c f) = Ordering c
+  ordering x y z o = bool (bool x y (o == eq)) z (o == gt)
+  compare = bitwiseCompare `on` getBitsBE
+
+bitwiseCompare :: forall c . Symbolic c => c [] -> c [] -> Ordering c
+bitwiseCompare x y = fold ((zipWith (compare `on` Bool) `on` unpacked) x y)
+
+getBitsBE :: forall c f . (Symbolic c, LayoutFunctor f) => c f -> c []
 -- ^ @getBitsBE x@ returns a list of circuits computing bits of @x@, eldest to
 -- youngest.
 getBitsBE x = symbolicF (arithmetize x Proxy)
```

Comment:
We don't need to `arithmetize x` because `x` is already a `c f`

Reply:
removed `arithmetize`